streamingccl: support DeleteRange in tenant stream replication

gh-casper opened the pull request:
This PR supports DeleteRange operation both in stream producers
that process DelRange from rangefeed and in stream ingestion
processors that ingest DelRange as SST into destination.

For version >= 22.2, SSTs can also contain range tombstones,
i.e., DelRange, this PR also supports parsing and ingesting them.

Ingestion processor uses a separate SST writer to ingest range
tombstones. Another PR will extend SSTBatcher to support
adding MVCCRangeKeys.

This PR creates an envelope type for RangeFeedSSTTable 
with additional field "RegisteredSpan" tracking the span of 
rangefeed registration this event is supposed to be matching. 
This helps rangefeed receiver to avoid processing rangefeed
events multiple times when it has multiple matching spans
registered.

This PR also cleans up the GenerationEvent that is no longer
a valid concept in the current consumer-tracked design.

Closes: https://github.com/cockroachdb/cockroach/issues/70433
Closes: https://github.com/cockroachdb/cockroach/issues/83810

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84553)
<!-- Reviewable:end -->

miretskiy commented on pkg/ccl/streamingccl/event.go:
```diff
@@ -95,66 +103,83 @@ func (sste sstableEvent) GetSSTable() *roachpb.RangeFeedSSTable {
 
 // GetKV implements the Event interface.
-func (ce checkpointEvent) GetKV() *roachpb.KeyValue {
+func (dre delRangeEvent) GetKV() *roachpb.KeyValue {
```
I think this really highlights the fact that having generic Event type is not helping.
every implementation has all but 1 method returning nil;

I think Event could be just a concrete type (a union).
Okay to leave a TODO for a cleanup later.

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -332,42 +342,58 @@ func (p *checkpointPacer) shouldCheckpoint(
-	sst *roachpb.RangeFeedSSTable, op func(keyVal roachpb.KeyValue),
+// Extract the received SST to only contain data within the boundaries of spans in the partition
+// spec, execute the specified operation on each MVCC key value in the trimmed SSTable. and
+// execute the specified operation on each MVCCRangeKey value in the trimmed SSTable.
```
nit: a bit better wrapping; also ". and" --> just start new stense.

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -82,14 +86,26 @@ type streamIngestionProcessor struct {
+	rangeKeySstFile *storage.MemFile
+	// Separated SSTWriter to write MVCC range keys and will be replaced after
+	// SSTBatcher supports adding MVCCRangeKeys.
+	rangeKeySSTWriter storage.SSTWriter
```
would be nice to split those range related fields into separate struct
rangeState struct {
   ....
}
or similar.
Also would be nice to have a pretty substantial comment the mechanics of how that data should be flushed; frontier updates, etc.

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -243,6 +260,9 @@ func (sip *streamIngestionProcessor) Start(ctx context.Context) {
 
+	sip.rangeKeySstFile = &storage.MemFile{}
+	sip.rangeKeySSTWriter = storage.MakeIngestionSSTWriter(ctx, evalCtx.Settings, sip.rangeKeySstFile)
+
```
do we want to create this lazily?

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -462,6 +482,19 @@ func (sip *streamIngestionProcessor) merge(
+		MVCCValueHeader: enginepb.MVCCValueHeader{LocalTimestamp: hlc.ClockTimestamp{WallTime: 0}},
+	})
+	if err != nil {
+		panic(err)
```
can we fail permanently instead?

erikgrinaker commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -561,30 +595,86 @@ func (sip *streamIngestionProcessor) bufferSST(sst *roachpb.RangeFeedSSTable) er
+			}); err != nil {
+				return err
+			}
+		} else if hasRange {
```
This isn't strictly correct, since it's possible for a range key and point key to be initially surfaced together. This can only happen for point keys without timestamps, i.e. intents or inline values, where the point key and range key's start bound are identical. Normally, we handle this by caching and comparing `RangeBounds().Key` (until #84379).

However, we can do something far more efficient here, by processing point and range keys separately, since we ingest them separately anyway. That is, first use `IterKeyTypePointsOnly` to process the point keys, then `IterKeyTypeRangesOnly` to process the range keys.

erikgrinaker commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -332,42 +342,58 @@ func (p *checkpointPacer) shouldCheckpoint(
+		}
+		if hasPoint, hasRange := iter.HasPointAndRange(); hasPoint {
+			mvccKeyValOp(iter.Key(), iter.Value())
+		} else if hasRange {
```
As mentioned in another comment, let's iterate over point keys and range keys separately, unless we need to emit them interleaved in order (in which case we need to detect changes to `RangeBounds().Key`).

erikgrinaker commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -385,10 +416,45 @@ func (s *eventStream) addSST(
+	mergeDelRange := func(unsafeRangeKey storage.MVCCRangeKey) {
+		delRanges, ok := timestampToDelRanges[unsafeRangeKey.Timestamp]
+		if ok {
+			for _, delRange := range delRanges {
```
Since we iterate across and add range keys in order (left-to-right in the key space), I believe we can just check the last member's end key here rather than iterating over all of them, which avoids making this quadratic.

erikgrinaker commented on pkg/roachpb/api.proto:
```diff
@@ -2793,14 +2795,17 @@ message RangeFeedSSTable {
   bytes              data     = 1;
   Span               span     = 2 [(gogoproto.nullable) = false];
   util.hlc.Timestamp write_ts = 3 [(gogoproto.nullable) = false, (gogoproto.customname) = "WriteTS"];
+  Span               registered_span     = 4 [(gogoproto.nullable) = false];
```
This seems a bit iffy. Shouldn't the client know this and keep track of it? Is it because the DistSender hides this internally?

Also, I see why we might need this on the SST (which isn't trimmed), but the DeleteRange should be truncated to the registration bounds already, so I don't think it should be necessary there in any case?

erikgrinaker reviewed (commented):
Some high-level comments around range tombstones, leaving the detailed review to those more familiar with tenant streaming.

gh-casper reviewed (commented):
Thanks! Resolved Yevgeniy's comments as the first batch. Will work on Erik's on the next batch.

There are some flakiness in the added tests primarily due to the test setup. I'll keep digging into those while in review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-:-N7N0GJTAy0JBjHERm11:bz9ure4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/event.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXd2cyuuwwUCh5N-r1-129:-N7MOVxZ-xdIb1_O5N5p:bjtpzbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/event.go#L129)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think this really highlights the fact that having generic Event type is not helping.
every implementation has all but 1 method returning nil;

I think Event could be just a concrete type (a union).
Okay to leave a TODO for a cleanup later.
</blockquote></details>

Added a todo here.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r1-108:-N7MwG907H4XK5A8D4DZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L108)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

would be nice to split those range related fields into separate struct
rangeState struct {
   ....
}
or similar.
Also would be nice to have a pretty substantial comment the mechanics of how that data should be flushed; frontier updates, etc.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r1-265:-N7Mqq7D0PYdVy_XPtvZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L265)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do we want to create this lazily? 
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 490 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r1-490:-N7MyZnx1JZhIm5Y6nbV:b1bdoxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L490)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

can we fail permanently instead?
</blockquote></details>

Done. Also have done some refactoring.



_Code quote:_
```Go
rangeTombstone
```


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 623 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r1-623:-N7N82lPCTK37a-L6IMx:bkmmlsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L623)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This isn't strictly correct, since it's possible for a range key and point key to be initially surfaced together. This can only happen for point keys without timestamps, i.e. intents or inline values, where the point key and range key's start bound are identical. Normally, we handle this by caching and comparing `RangeBounds().Key` (until #84379).

However, we can do something far more efficient here, by processing point and range keys separately, since we ingest them separately anyway. That is, first use `IterKeyTypePointsOnly` to process the point keys, then `IterKeyTypeRangesOnly` to process the range keys.
</blockquote></details>

What does `RangeBounds` mean? It returns the boundary of all of current `RangeKeys()`? 
`
RangeKeys() -> ["a", "c"], ["b", "d"], 
RangeBounds() -> ["a", "d"]
`
___
*[`pkg/roachpb/api.proto` line 2798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRu-r1-2798:-N7N5eOp8Y_XkU05TxvP:b-x4rnxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/roachpb/api.proto#L2798)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This seems a bit iffy. Shouldn't the client know this and keep track of it? Is it because the DistSender hides this internally?

Also, I see why we might need this on the SST (which isn't trimmed), but the DeleteRange should be truncated to the registration bounds already, so I don't think it should be necessary there in any case?
</blockquote></details>

If the client creates rangefeed with two spans.
`
rf := RangefeedFactory.New(...)
rf.Start(ctx, []roachpb.Span{"1", "2"})
`

Whenever a SST that overlaps with both "1" and "2", a same `*RangefeedSSTable` event will be emitted twice as there are two registrations. I want to avoid this. Have considered using if the current event is the same pointer as the last received one. This is hacky and may not be correct, as this event may not be sent consecutively.

So I added `RegisteredSpan` here to let client know which overlapping span this event is about. You may say what if client start the rangefeed with two same spans, but that's client's problem at the first place.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 623 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r1-623:-N7QB-c9CUsBzlLtpJCq:b-g1u3gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L623)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

What does `RangeBounds` mean? It returns the boundary of all of current `RangeKeys()`? 
`
RangeKeys() -> ["a", "c"], ["b", "d"], 
RangeBounds() -> ["a", "d"]
`
</blockquote></details>

Because of fragmentation, all range keys at a given position will all have the same bounds. See: https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/mvcc-range-tombstones.md#data-structure-mvcc-range-keys

So for your example of overlapping range keys `[a-c)` and `[b-d)`, these would get fragmented into `[a-b)`, `[b-c)`, and `[c-d)`, and you would only see the ones overlapping the iterator position.
___
*[`pkg/roachpb/api.proto` line 2798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRu-r1-2798:-N7QBwioCzfSjkgnixyy:b-xze3hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/roachpb/api.proto#L2798)):*
> Have considered using if the current event is the same pointer as the last received one. This is hacky and may not be correct, as this event may not be sent consecutively.

Yeah, I agree, we definitely shouldn't do that.

I suppose this is fine. Since it's the caller's responsibility to trim the SST, we may as well make it easy for them to do so. However, we should:

* Mention in the comment that the caller should use this to trim the contents (as outlined in the comment below).
* Add an assertion in `TestReplicaRangefeed`.
* Add the SST span and registration span to `kvclient/rangefeed/OnSSTable()`, and test it in `TestWithOnSSTable`.

We should remove this for `DeleteRange` though, since that's always truncated to the registration span.

Are there any mixed-version compatibility concerns here? In other words, is it possible to use this streaming in a cluster that hasn't yet completed an upgrade to 22.2, where 22.1 nodes will not be setting this field? If so, we'll need a version gate here. We should also mention in the comment that this field is new in 22.2.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/roachpb/api.proto` line 2798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRu-r1-2798:-N7QZxKN0n1wpDMvdg_M:b3de4uy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/roachpb/api.proto#L2798)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> Have considered using if the current event is the same pointer as the last received one. This is hacky and may not be correct, as this event may not be sent consecutively.

Yeah, I agree, we definitely shouldn't do that.

I suppose this is fine. Since it's the caller's responsibility to trim the SST, we may as well make it easy for them to do so. However, we should:

* Mention in the comment that the caller should use this to trim the contents (as outlined in the comment below).
* Add an assertion in `TestReplicaRangefeed`.
* Add the SST span and registration span to `kvclient/rangefeed/OnSSTable()`, and test it in `TestWithOnSSTable`.

We should remove this for `DeleteRange` though, since that's always truncated to the registration span.

Are there any mixed-version compatibility concerns here? In other words, is it possible to use this streaming in a cluster that hasn't yet completed an upgrade to 22.2, where 22.1 nodes will not be setting this field? If so, we'll need a version gate here. We should also mention in the comment that this field is new in 22.2.
</blockquote></details>

What I don't understand is why we need this at the API level.
When starting rangefeed, dist sender rangefeed knows exactly the bounds it's dealing with (partial rangefeed).
Couldn't we extend partial rangefeed to return bounds? I realize right now it's a bit clunkly because single rangefeed just does 
```

			select {
			case eventCh <- event:
			case <-ctx.Done():
				return args.Timestamp, ctx.Err()
			}
```
where eventCh is a *RangeFeedEvent channel, with event arriving from rangefeed directly; 
but I can't imagine this being *that* hard to make the change in the client API, or perhaps introduce another method that can take a different kind of channel, and have this information available on the client side?

Wdyt @erikgrinaker ?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/roachpb/api.proto` line 2798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRu-r1-2798:-N7QeaAVBoIFsmKBSe24:bu6ri0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/roachpb/api.proto#L2798)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

What I don't understand is why we need this at the API level.
When starting rangefeed, dist sender rangefeed knows exactly the bounds it's dealing with (partial rangefeed).
Couldn't we extend partial rangefeed to return bounds? I realize right now it's a bit clunkly because single rangefeed just does 
```

			select {
			case eventCh <- event:
			case <-ctx.Done():
				return args.Timestamp, ctx.Err()
			}
```
where eventCh is a *RangeFeedEvent channel, with event arriving from rangefeed directly; 
but I can't imagine this being *that* hard to make the change in the client API, or perhaps introduce another method that can take a different kind of channel, and have this information available on the client side?

Wdyt @erikgrinaker ?
</blockquote></details>

I was thinking the same thing, but then we'd have to introduce additional data structures on the client side to track it and such. It just seemed excessive, especially given that this is limited to a single event type that is a bit of an outlier.

Since exceeding the registration bounds is a property of this one single event, it seems reasonable to also include the relevant bounds in the event.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> I feel like we're just adding on cruft for the benefit of a very specialized use-case though.

So, would that cruft be better on kv server side? I think that's the question -- and my preference is to ... not add cruft on rangefeed server?
Wdyt?

gh-casper reviewed (commented):
Okay, let me try message envelope approach, will have a separate commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-:-N7RDv0ADsv07UKKmXMA:b-ab4jz9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

gh-casper committed:
kv: create RangeFeedMessage type to encapsulate RangeFeedEvent

This PR creates kvcoor.RangeFeedMessage encapsulating
RangeFeedEvent and additional RegisteredSpan field which
is the span of the rangefeed registration that emits this
event.

This field is to help rangefeed clients to properly trim
the received SST with only single registered span. In the
case of creating rangefeed with multiple spans, rangefeed
consumers may end up receiving duplicate SST without
proper trimming using this field.

Release note: None

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 623 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r1-623:-N7VMlV-DB02vgL0fFFf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L623)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Because of fragmentation, all range keys at a given position will all have the same bounds. See: https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/mvcc-range-tombstones.md#data-structure-mvcc-range-keys

So for your example of overlapping range keys `[a-c)` and `[b-d)`, these would get fragmented into `[a-b)`, `[b-c)`, and `[c-d)`, and you would only see the ones overlapping the iterator position.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRq-r1-375:-N7VMtigE8roHpHlxZu2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamproducer/event_stream.go#L375)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

As mentioned in another comment, let's iterate over point keys and range keys separately, unless we need to emit them interleaved in order (in which case we need to detect changes to `RangeBounds().Key`).
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 425 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRq-r1-425:-N7VMvJzDDjuEnPDwMuZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/ccl/streamingccl/streamproducer/event_stream.go#L425)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Since we iterate across and add range keys in order (left-to-right in the key space), I believe we can just check the last member's end key here rather than iterating over all of them, which avoids making this quadratic.
</blockquote></details>

Done.
___
*[`pkg/roachpb/api.proto` line 2798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRu-r1-2798:-N7VMhq1-K7Hk_EINurC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab49172f901ea4c23431cd9bd3473140cae568b/pkg/roachpb/api.proto#L2798)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I was thinking the same thing, but then we'd have to introduce additional data structures on the client side to track it and such. It just seemed excessive, especially given that this is limited to a single event type that is a bit of an outlier.

Since exceeding the registration bounds is a property of this one single event, it seems reasonable to also include the relevant bounds in the event.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
Nice, range key and rangefeed parts look good at a high level! I'll leave the detailed review of the tenant streaming parts to that team.

> Reviewed 30 of 30 files at r3, 18 of 18 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-:-N7r4s05BfA-S8bJyWqd:bxjo53q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/utils.go` line 63 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7r0KmV8zMEgDChEkOq:-N7r0KmV8zMEgDChEkOr:b-wufuzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2a64a8be2875e5df788558e08edb953b7ece867/pkg/ccl/streamingccl/utils.go#L63)):*
> ```Go
> 			break
> 		}
> 		if hasPoint, _ := pointIter.HasPointAndRange(); hasPoint {
> ```

We don't need to check this. With `IterKeyTypePointsOnly`, this will always be `true` for a valid iterator. Ditto for `IterKeyTypeRangesOnly`.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 750 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7r1oEB5eQdYe5AO9uM:-N7r1oEB5eQdYe5AO9uN:b-vx48t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2a64a8be2875e5df788558e08edb953b7ece867/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L750)):*
> ```Go
> 			return 0, err
> 		}
> 		if err = r.rangeKeySSTWriter.PutMVCCRangeKey(rangeKeyVal.RangeKey, mvccValue); err != nil {
> ```

I have an open PR for `PutRawMVCCRangeKey()`, which will avoid having to decode/encode the value when copying like this.

#84962

We should make sure we update this once it's merged.
___
*[`pkg/kv/kvclient/rangefeed_envelope.go` line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7qkkb13iVw2yKdhSra:-N7qkkb13iVw2yKdhSrb:b-vjgqcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/64a584e875b04e452043fe7b1bbd9facbaafcb26/pkg/kv/kvclient/rangefeed_envelope.go#L16)):*
> ```Go
> 
> // RangeFeedEnvelope is envelope type that encapsulates the roachpb.RangeFeedEvent.
> type RangeFeedEnvelope struct {
> ```

I don't love the "envelope" name, but don't have any great other ideas either. Maybe `Message`?  What do you think?

Also, we should probably move this into `kvcoord`, which is the package that constructs and emits them via `DistSender`.
___
*[`pkg/kv/kvclient/rangefeed_envelope.go` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7r4SIk4YAoLWk7Rvej:-N7r4SIk4YAoLWk7Rvek:bvyaha5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2a64a8be2875e5df788558e08edb953b7ece867/pkg/kv/kvclient/rangefeed_envelope.go#L23)):*
> ```Go
> 	// The span of the rangefeed registration that overlaps with SST span if
> 	// event has an underlying roachpb.RangeFeedSSTable event.
> 	RegisteredSpan *roachpb.Span
> ```

This can be a value rather than a pointer.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7ql6c-8dwAHyw77zyX:-N7ql6c-8dwAHyw77zyY:b5oqnpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/64a584e875b04e452043fe7b1bbd9facbaafcb26/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L509)):*
> ```Go
> 				}
> 			case *roachpb.RangeFeedSSTable:
> 				envelope.RegisteredSpan = &span
> ```

Shouldn't we always set this regardless of event type?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 520 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7qya8B-UpfGcHKIG2N:-N7qya8B-UpfGcHKIG2O:bisk6y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/64a584e875b04e452043fe7b1bbd9facbaafcb26/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L520)):*
> ```Go
> 
> 			select {
> 			case eventCh <- &envelope:
> ```

We should consider sending a value rather than a pointer, which avoids it escaping to the heap. Ran some quick benchmarks which just built and sent the structs over a buffered channel:

```
BenchmarkChannelEvent-24            	 6036284	       197.3 ns/op
BenchmarkChannelEnvelope-24         	 5102538	       235.1 ns/op
BenchmarkChannelEnvelopeValue-24    	 5612910	       217.3 ns/op
```

So sending it by value cuts about half of the overhead of the envelope.
___
*[`pkg/kv/kvserver/replica_rangefeed_test.go` line 380 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7r-V5A4bBf6R3ge2si:-N7r-V5A4bBf6R3ge2sj:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/64a584e875b04e452043fe7b1bbd9facbaafcb26/pkg/kv/kvserver/replica_rangefeed_test.go#L380)):*
> ```Go
> 	require.NoError(t, sstWriter.Finish())
> 
> 	// Does this mean all ts in sst should be equal to the batch ts?
> ```

Yes.


<!-- Sent from Reviewable.io -->

gh-casper committed:
streamingccl: support DeleteRange in tenant stream replication

This PR supports DeleteRange operation both in producer that
process DelRange from rangefeed and in stream ingestion
processor that ingest DelRange as SST into destination.

For version >= 22.2, SSTs can also contain range tombstones,
i.e., DelRange, this PR also supports ingesting them.

Ingestion processor uses a separate SST writer to ingest range
tombstones as SSTBatcher does not support adding MVCCRangeKeys.
yet.

This PR also cleans up the GenerationEvent that is no longer
a valid concept in the current consumer-tracked design.

Release note: None

gh-casper reviewed (commented):
Thanks Erik. Please review for the streaming part. @stevendanna @dt @miretskiy 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-:-N80ZwyOCGh_lPKe2uoc:bh0o4ka)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/kv/kvclient/kvcoord/rangefeed_message.go:
```diff
@@ -0,0 +1,24 @@
+	*roachpb.RangeFeedEvent
+
+	// The span of the rangefeed registration that overlaps with SST span if
+	// event has an underlying roachpb.RangeFeedSSTable event.
```
I wouldn't mention anything about SST;
Isn't this span a "range" against which rangefeed was established?

erikgrinaker commented on pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:
```diff
@@ -87,7 +88,7 @@ func (ds *DistSender) RangeFeed(
 	startAfter hlc.Timestamp, // exclusive
 	withDiff bool,
-	eventCh chan<- *roachpb.RangeFeedEvent,
+	eventCh chan<- *kvclient.RangeFeedEnvelope,
```
Wouldn't we then have to migrate all callers over to this new `ConsumeRangeFeed` method anyway, which is presumably an even larger change since it'd introduce callbacks everywhere?

I don't have a strong opinion on channels vs callbacks -- slight preference for channels because they can be combined via `select`. If you generally prefer callbacks on the client side then that's fine with me, since CDC is the primary consumer anyway.

I'm not sure if it's the best use of @gh-casper's time for him to keep rewriting this though. If he'd rather keep it as-is and work on something more important then I'd be +1 on that.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -643,37 +731,127 @@ func (sip *streamIngestionProcessor) bufferCheckpoint(event partitionEvent) erro
+		if err != nil {
+			return 0, err
+		}
+		if err = r.rangeKeySSTWriter.PutMVCCRangeKey(rangeKeyVal.RangeKey, mvccValue); err != nil {
```
Looks like you might be able to use `PutRawMVCCRangeKey` now to avoid the decode.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -643,37 +731,127 @@ func (sip *streamIngestionProcessor) bufferCheckpoint(event partitionEvent) erro
+		r.rangeKeySSTWriter = r.rangeKeySSTWriterMaker()
+	}
+
+	for _, rangeKeyVal := range rangeKeys {
```
Are there ordering constraints on how we write range keys into the SST? That is, do we need sort this like we sort KV keys?

erikgrinaker replied:
Yes, they must be written in key order. Multiple versions of the same range key can be written in any order, as long as the key order is maintained.

stevendanna reviewed (commented):
Overall this looks like a pretty reasonable step to me. I have one question about whether we need to sort the range keys.

Overall, I think we probably want some memory monitoring here just like we'll want some for the existing kv batching, but we can do that in a follow up.

gh-casper reviewed (commented):
Memory monitoring in ingestion is something that I plan to do in the August plan, either through BufferAdder or set up its own memory monitoring mechanism.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-:-N8TzOxF2jqNUe06O-c4:bstl8zs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r9-133:-N8U0xtB0uRNQUnot6YI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b014f6ffc8ae730ba4d9df63edc523478b9377da/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L133)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should this be `'rangeBatcher' below`?
</blockquote></details>

Done.



_Code quote:_
```Go
curRangeKVBatch
```
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 667 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r9-667:-N8U25Uq5XDz6yLXlQIR:b-mjewyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b014f6ffc8ae730ba4d9df63edc523478b9377da/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L667)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] should we use this function down below in `bufferKV` as well?
</blockquote></details>

right, good call.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 745 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r9-745:-N8ZdWGCEVw_Nn7OVmQu:b-ilpdgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b014f6ffc8ae730ba4d9df63edc523478b9377da/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L745)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yes, they must be written in key order. Multiple versions of the same range key can be written in any order though, as long as the key order is maintained.
</blockquote></details>

Now I sort them before batching. Also refactored the rangeBatcher to make it cleaner.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go` line 750 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRk-r9-750:-N8U2ahz0pz51cpv8fOF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b014f6ffc8ae730ba4d9df63edc523478b9377da/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L750)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Looks like you might be able to use `PutRawMVCCRangeKey` now to avoid the decode.
</blockquote></details>

Done.



_Code quote:_
```Go
PutMVCCRangeKey
```
___
*[`pkg/kv/kvclient/kvcoord/rangefeed_message.go` line 22 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N80Zc-pxNwtzBszUSZC-r7-22:-N8U34t09XHq7XJwhWtN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e8c53c5df9fc0fe9f97300b8e5b3eaad9252d37/pkg/kv/kvclient/kvcoord/rangefeed_message.go#L22)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wouldn't mention anything about SST;
Isn't this span a "range" against which rangefeed was established?
</blockquote></details>

Done.



_Code quote:_
```Go
has an underlying roachpb.RangeFeedSSTable event.
```


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:
```diff
@@ -324,6 +325,10 @@ func TestTenantStreamingSuccessfulIngestion(t *testing.T) {
 
+	// TODO(casper): disabled due to error when setting a cluster setting
+	// "setting updated but timed out waiting to read new value"
+	skip.UnderStressRace(t, "disabled under stress race")
```
Let's open an issue for this so we don't forget.

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 330 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84553#-N7EDKXeaFt63_6rfWRm-r11-330:-N8dEKxAAtDEmUm630Va:bfy2anj) ([raw file](https://github.com/cockroachdb/cockroach/blob/dae630410dee5c06a4d38f7790cb754686c5f1e2/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L330)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Let's open an issue for this so we don't forget.
</blockquote></details>

Done. https://github.com/cockroachdb/cockroach/issues/85615



_Code quote:_
```Go
	// TODO(casper): disabled due to error when setting a cluster setting
	// "setting updated but timed out waiting to read new value"
```


<!-- Sent from Reviewable.io -->

gh-casper commented:
bors r+
